### PR TITLE
PMM-9946 Automatically add k8s cluster

### PR DESCRIPTION
**Pull request by gen1us2k**
* Check if PMM is inside k8s cluster
* Get config and convert it to kubeconfig
* Trigger register cluster call

PMM-0

Build: SUBMODULES-0

- [ ] Links to other linked pull requests (optional).

### CODE REVIEW

**Comment by gen1us2k** [comment]
There's no need for review at this moment. I'm just trying to create PoC of this automation

File: managed/services/dbaas/dbaas_client.go [hunk 1]
```diff
@@ -54,6 +55,30 @@ func NewClient(dbaasControllerAPIAddress string) *Client {
...
+	}
+	c.k8sClient = client
+	return nil
+}
+func (c *Client) GetKubeConfig() (string, error) {
+
+	if err := c.initializeInClusterK8sClient(); err != nil {
+		return "", err
+	}
+	if err := c.Connect(context.Background()); err != nil {
```
Comment (gen1us2k) [inline managed/services/dbaas/dbaas_client.go:71]: @BupycHuk I don't like connecting to dbaas-controller here. It kind of breaks logic of enabling/disabling dbaas feature. Any suggestions here?

File: managed/services/dbaas/kubernetes_client.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
+package dbaas
+
+import (
+	"context"
+	"encoding/json"
+
```
Comment (gen1us2k) [inline managed/services/dbaas/kubernetes_client.go:6]: Tests are missing here. I'll fix it soon

File: managed/services/dbaas/dbaas_client.go (hunk 1 above)
Comment (BupycHuk) [reply managed/services/dbaas/dbaas_client.go:71]: in the case of disabled dbaas, we should return an empty value by default.

File: managed/services/management/dbaas/dbaas_initializer.go [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+// Enable enables dbaas feature and connects to dbaas-controller
+func (in *Initializer) Enable(ctx context.Context) error {
+	in.m.Lock()
+	defer in.m.Unlock()
+	if in.enabled {
+		return nil
+	}
+	timeoutCtx, cancel := context.WithTimeout(ctx, 20*time.Second)
+	err := in.dbaasClient.Connect(timeoutCtx)
+	cancel()
```
Comment (oksana-grishchenko) [inline managed/services/management/dbaas/dbaas_initializer.go:90]: shouldn't it be 
```
defer cancel()
``` 
here?

File: managed/services/management/dbaas/dbaas_initializer.go [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+		req := &dbaasv1beta1.RegisterKubernetesClusterRequest{
+			KubernetesClusterName: defaultClusterName,
+			KubeAuth: &dbaasv1beta1.KubeAuth{
+				Kubeconfig: kubeConfig.Kubeconfig,
+			},
+		}
+		_, err = in.RegisterCluster(ctx, req)
+		if err != nil {
+			return err
+		}
```
Comment (oksana-grishchenko) [inline managed/services/management/dbaas/dbaas_initializer.go:127]: Suggestion: add logs before all `return`s to simplify troubleshooting

File: managed/services/management/dbaas/dbaas_initializer.go [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+	}
+	pmmVersion, err := goversion.NewVersion(pmmversion.PMMVersion)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	pxcOperatorVersion, psmdbOperatorVersion, err := in.versionService.LatestOperatorVersion(ctx, pmmVersion.Core().String())
+	if err != nil {
+		return nil, err
+	}
```
Comment (oksana-grishchenko) [inline managed/services/management/dbaas/dbaas_initializer.go:192]: Suggestion: add logs before all `return`s to simplify troubleshooting

**Review by oksana-grishchenko — APPROVED** [review]
Posted some comments, everything else LGTM

File: managed/services/management/dbaas/dbaas_initializer.go [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+	timeoutCtx, cancel := context.WithTimeout(ctx, 20*time.Second)
+	err := in.dbaasClient.Connect(timeoutCtx)
+	cancel()
+	if err != nil {
+		return err
+	}
+	ctx, in.cancel = context.WithCancel(ctx)
+
+	in.enabled = true
+	kubeConfig, err := in.dbaasClient.GetKubeConfig(ctx, &dbaascontrollerv1beta1.GetKubeconfigRequest{})
```
Comment (BupycHuk) [inline managed/services/management/dbaas/dbaas_initializer.go:97]: can we extract everything below this line to a separate method?

Comment (BupycHuk) [reply managed/services/management/dbaas/dbaas_initializer.go:97]: And should we do it each time?

File: managed/services/management/dbaas/dbaas_initializer.go [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+	err := in.dbaasClient.Disconnect()
+	if err != nil {
+		return err
+	}
+	in.enabled = false
+	return nil
+}
+
+// RegisterCluster registers k8s cluster and installs all required operators
+func (in *Initializer) RegisterCluster(ctx context.Context, req *dbaasv1beta1.RegisterKubernetesClusterRequest) (*dbaasv1beta1.RegisterKubernetesClusterResponse, error) {
```
Comment (BupycHuk) [inline managed/services/management/dbaas/dbaas_initializer.go:155]: TBH I don't like the idea that we moved this method here. It doesn't feel like it belongs here.

**Review by BupycHuk — CHANGES_REQUESTED** [review]

File: managed/services/management/dbaas/dbaas_initializer.go (hunk 3 above)
Comment (gen1us2k) [reply managed/services/management/dbaas/dbaas_initializer.go:90]: Yep. Thanks

File: managed/services/management/dbaas/dbaas_initializer.go (hunk 7 above)
Comment (gen1us2k) [reply managed/services/management/dbaas/dbaas_initializer.go:162]: What do you suggest?

File: managed/services/management/dbaas/dbaas_initializer.go (hunk 6 above)
Comment (gen1us2k) [reply managed/services/management/dbaas/dbaas_initializer.go:102]: I moved it to a separate function. 

> And should we do it each time?

Yes, but we do not do anything else once PMM has registered cluster.

File: managed/services/management/dbaas/dbaas_initializer.go (hunk 7 above)
Comment (BupycHuk) [reply managed/services/management/dbaas/dbaas_initializer.go:162]: 1. The best option would be to have layered architecture, so we wouldn't call a gRPC server. But it makes the code inconsistent with what we have in the project now.
2. Copy and paste?
3. Calling kubernetes_server from the initializer.

**Review by BupycHuk — APPROVED** [review]

File: managed/services/management/dbaas/dbaas_initializer.go [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+			if err != nil {
+				in.l.Errorf("failed finding cluster: %v", err)
+				return nil
+			}
+			if cluster != nil {
+				return errClusterExists
+			}
+			return nil
+		})
+		if err != nil {
```
Comment (BupycHuk) [inline managed/services/management/dbaas/dbaas_initializer.go:111]: err can be only `errClusterExists`

File: managed/services/management/dbaas/dbaas_initializer.go [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+				Kubeconfig: kubeConfig.Kubeconfig,
+			},
+		}
+		_, err = in.kubernetesServer.RegisterKubernetesCluster(ctx, req)
+		if err != nil {
+			return err
+		}
+		in.l.Info("Cluster is successfully initialized")
+	} else {
+		in.l.Errorf("failed getting kubeconfig inside cluster: %v", err)
```
Comment (BupycHuk) [inline managed/services/management/dbaas/dbaas_initializer.go:130]: what response do we get if we run PMM outside of the cluster?

**Pull request merged**